### PR TITLE
Update persistent DBUS objects from Host

### PR DESCRIPTION
**Pull request by lxwinspur**
This commit destroy & create new core DBus objects for every
first getPDR command.

Refer https://github.com/ibm-openbmc/dev/issues/3547

Signed-off-by: George Liu <liuxiwei@inspur.com>

### CODE REVIEW

File: host-bmc/host_pdr_handler.cpp [hunk 1]
```diff
@@ -840,6 +840,15 @@ void HostPDRHandler::processHostPDRs(mctp_eid_t /*eid*/,
         pldm::hostbmc::utils::updateEntityAssociation(
             entityAssociations, entityTree, objPathMap, oemPlatformHandler);
 
+        if (this->isfirstgetPDR)
```
Comment (manojkiraneda) [inline host-bmc/host_pdr_handler.cpp:843]: this logic should be moved completely to the getPDR function in platform.cpp right ? - we need to do this before even getting the PDR's from host (not when the last record handle from host is null)

File: host-bmc/host_pdr_handler.cpp [hunk 2]
```diff
@@ -840,6 +840,15 @@ void HostPDRHandler::processHostPDRs(mctp_eid_t /*eid*/,
...
+        {
+            this->isfirstgetPDR = false;
+            pldm::dbus::removeCore();
```
[the remark this answers is not in the document]
Comment (lxwinspur) [reply host-bmc/host_pdr_handler.cpp]: > `pldm::dbus::CustomDBus::getCustomDBus().removeCore()`

Done

File: host-bmc/dbus/serialize.cpp [hunk 3]
```diff
@@ -100,5 +100,31 @@ void Serialize::setObjectPathMaps(const ObjectPathMaps& maps)
     }
 }
 
+void reSerialize(const std::vector<uint16_t> types)
```
[the remark this answers is not in the document]
Comment (lxwinspur) [reply host-bmc/dbus/serialize.cpp]: > `void Serialize::reSerialize(const std::vector<uint16_t> types)`

Done

File: host-bmc/host_pdr_handler.cpp (hunk 1 above)
Comment (lxwinspur) [reply host-bmc/host_pdr_handler.cpp]: > this logic should be moved completely to the getPDR function in platform.cpp right ? - we need to do this before even getting the PDR's from host (not when the last record handle from host is null)

Ack

File: host-bmc/dbus/custom_dbus.cpp [hunk 4]
```diff
@@ -331,5 +334,130 @@ void CustomDBus::setMicrocode(const std::string& path, uint32_t value)
...
+    }
+
+    auto savedObjs = pldm::serialize::Serialize::getSerialize().getSavedObjs();
+    for (const auto& type : types)
+    {
+        if (!savedObjs.contains(type))
+        {
+            continue;
+        }
+
```
Comment (manojkiraneda) [inline host-bmc/dbus/custom_dbus.cpp:454]: Add below trace here :
`std::cerr << "Deleting the dbus objects of type : " << (unsigned)type << std::endl;`

File: host-bmc/dbus/serialize.cpp [hunk 5]
```diff
@@ -100,5 +100,31 @@ void Serialize::setObjectPathMaps(const ObjectPathMaps& maps)
...
+    if (types.empty())
+    {
+        return;
+    }
+
+    for (const auto& type : types)
+    {
+        if (savedObjs.contains(type))
+        {
+            savedObjs.erase(savedObjs.find(type));
```
Comment (manojkiraneda) [inline host-bmc/dbus/serialize.cpp:114]: before erase, please add this trace :
`std::cerr << "Removing objects of type : " << (unsigned)type << " from the persistent cache\n";`

**Review by manojkiraneda — COMMENTED** [review]
Please add few traces that i mentioned, apart from that looks good to me.

File: host-bmc/dbus/serialize.cpp (hunk 5 above)
Comment (lxwinspur) [reply host-bmc/dbus/serialize.cpp:116]: > std::cerr << "Removing objects of type : " << (unsigned)type << " from the persistent cache\n";

Ack

File: host-bmc/dbus/custom_dbus.cpp (hunk 4 above)
Comment (lxwinspur) [reply host-bmc/dbus/custom_dbus.cpp:456]: > Add below trace here : `std::cerr << "Deleting the dbus objects of type : " << (unsigned)type << std::endl;`

Ack

**Comment by manojkiraneda** [comment]
Can one of the admins verify this patch?

**Review by manojkiraneda — APPROVED** [review]

**Review by sampmisr — APPROVED** [review]

**Comment by manojkiraneda** [comment]
fixes : [SW548423](https://w3.rchland.ibm.com/projects/bestquest/?defect=SW548423)

**Comment by rfrandse** [comment]
superseded by https://github.com/ibm-openbmc/pldm/pull/214

**Pull request closed without merging**